### PR TITLE
Added Skellam distribution

### PR DESCRIPTION
**Pull request by ritesh99rakesh**
#### References to other Issues or PRs

N/A

#### Brief description of what is fixed or changed
Added skellam distribution to `sympy/stats/drv_types.py`

#### Other comments
Ping @sidhantnagpal 

#### Release Notes

- stats
    - Added skellam distribution to `sympy/stats/drv_types.py`

### CODE REVIEW

File: sympy/stats/drv_types.py [hunk 1]
```diff
@@ -367,6 +368,97 @@ def Poisson(name, lamda):
...
+    def _cdf(self, x):
+        return NotImplementedError("Skellam doesn't have closed form for the CDF.")
+
+    def _characteristic_function(self, t):
+        mu1, mu2 = self.mu1, self.mu2
+        return exp(-(mu1+mu2)+mu1*exp(t)+mu2*exp(-t))
+
+    def _moment_generating_function(self, t):
+        mu1, mu2 = self.mu1, self.mu2
+        return exp(-(mu1+mu2)+mu1*exp(I*t)+mu2*exp(-I*t))
```
Comment (sidhantnagpal) [inline sympy/stats/drv_types.py]: Spaces around `+` (here and above).

File: sympy/stats/drv_types.py [hunk 2]
```diff
@@ -367,6 +368,97 @@ def Poisson(name, lamda):
...
+         z
+         -
+         2
+    /mu1\   -mu1 - mu2        /       _____   _____\
+    |---| *e          *besseli\z, 2*\/ mu1 *\/ mu2 /
+    \mu2/
+
+    >>> E(X)
+    I*mu1 - I*mu2
+
```
Comment (sidhantnagpal) [inline sympy/stats/drv_types.py]: This newline can be removed (else it will render separate code blocks).

File: sympy/stats/drv_types.py [hunk 3]
```diff
@@ -367,6 +368,97 @@ def Poisson(name, lamda):
...
+    >>> from sympy.stats import Skellam, density, E, variance
+    >>> from sympy import Symbol, simplify, pprint
+
+    >>> z = Symbol("z", integer=True)
+    >>> mu1 = Symbol("mu1", positive=True)
+    >>> mu2 = Symbol("mu2", positive=True)
+
+    >>> X = Skellam("x", mu1, mu2)
+
```
Comment (sidhantnagpal) [inline sympy/stats/drv_types.py]: Similar to https://github.com/sympy/sympy/pull/17099#discussion_r297739020.

File: sympy/stats/drv_types.py [hunk 4]
```diff
@@ -367,6 +368,97 @@ def Poisson(name, lamda):
...
+    I*mu1 - I*mu2
+
+    >>> simplify(variance(X))
+    -mu1 - mu2
+
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/Skellam_distribution
+    .. [2]
```
Comment (sidhantnagpal) [inline sympy/stats/drv_types.py]: Did you intend to add another reference?

**Comment by sidhantnagpal** [comment]
Can you add tests for mean, variance, skewness, etc?

File: sympy/stats/drv_types.py (hunk 4 above)
Comment (ritesh99rakesh) [reply sympy/stats/drv_types.py]: No

File: sympy/stats/drv_types.py [hunk 5]
```diff
@@ -367,6 +368,97 @@ def Poisson(name, lamda):
...
+    >>> pprint(density(X)(z), use_unicode=False)
+         z
+         -
+         2
+    /mu1\   -mu1 - mu2        /       _____   _____\
+    |---| *e          *besseli\z, 2*\/ mu1 *\/ mu2 /
+    \mu2/
+
+    >>> E(X)
+    I*mu1 - I*mu2
```
Comment (ethankward) [inline sympy/stats/drv_types.py]: The expectation should be real.

**Review by ethankward — CHANGES_REQUESTED** [review]

File: sympy/stats/drv_types.py (hunk 5 above)
Comment (ritesh99rakesh) [reply sympy/stats/drv_types.py]: Yes, I don't know why `I` is appearing here.

Comment (ethankward) [reply sympy/stats/drv_types.py]: It looks like you switched the moment generating / characteristic functions.

**Comment by ritesh99rakesh** [comment]
@sidhantnagpal Please review. Also, I have made some minor formatting changes using pep8 format.

File: sympy/stats/drv_types.py [hunk 6]
```diff
@@ -367,6 +368,94 @@ def Poisson(name, lamda):
...
+    >>> pprint(density(X)(z), use_unicode=False)
+         z
+         -
+         2
+    /mu1\   -mu1 - mu2        /       _____   _____\
+    |---| *e          *besseli\z, 2*\/ mu1 *\/ mu2 /
+    \mu2/
+    >>> E(X)
+    mu1 - mu2
+    >>> simplify(variance(X))
```
Comment (sidhantnagpal) [inline sympy/stats/drv_types.py]: `variance(X).expand()` should suffice here.

File: sympy/stats/tests/test_discrete_rv.py [hunk 7]
```diff
@@ -51,18 +59,39 @@ def test_negative_binomial():
...
+        exp(-mu1 - mu2)*besseli(z, 2*sqrt(mu1*mu2))
+    assert skewness(X).expand() == mu1/(mu1*sqrt(mu1 + mu2) + mu2 *
+                sqrt(mu1 + mu2)) - mu2/(mu1*sqrt(mu1 + mu2) + mu2*sqrt(mu1 + mu2))
+    assert variance(X).expand() == mu1 + mu2
+    assert E(X) == mu1 - mu2
+    assert characteristic_function(X)(z) == exp(
+        mu1*exp(I*z) - mu1 - mu2 + mu2*exp(-I*z))
+    assert moment_generating_function(X)(z) == exp(
+        mu1*exp(z) - mu1 - mu2 + mu2*exp(-z))
+
```
Comment (sidhantnagpal) [inline sympy/stats/tests/test_discrete_rv.py]: Test for `skewness`?

File: sympy/stats/drv_types.py (hunk 6 above)
Comment (ritesh99rakesh) [reply sympy/stats/drv_types.py]: @sidhantnagpal I will this change, but I believe that `simplify` is intuitive  and hence we should keep `simplify` in example code and use `.expand()` in tests as I have done.

File: sympy/stats/tests/test_discrete_rv.py (hunk 7 above)
Comment (ritesh99rakesh) [reply sympy/stats/tests/test_discrete_rv.py]: It is there on line 71

**Comment by ritesh99rakesh** [comment]
@sidhantnagpal Please review.

File: sympy/stats/tests/test_discrete_rv.py [hunk 8]
```diff
@@ -135,24 +168,31 @@ def test_moment_generating_functions():
...
 
-    negative_binomial_mgf = moment_generating_function(NegativeBinomial('n', 5, S(1)/3))(t)
+    negative_binomial_mgf = moment_generating_function(
+        NegativeBinomial('n', 5, S(1)/3))(t)
     assert negative_binomial_mgf.diff(t).subs(t, 0) == S(5)/2
 
     poisson_mgf = moment_generating_function(Poisson('p', 5))(t)
     assert poisson_mgf.diff(t).subs(t, 0) == 5
 
+    sellam_mgf = moment_generating_function(Skellam('s', 1, 1))(t)
```
Comment (sidhantnagpal) [inline sympy/stats/tests/test_discrete_rv.py]:
```suggestion
    skellam_mgf = moment_generating_function(Skellam('s', 1, 1))(t)
```

File: sympy/stats/tests/test_discrete_rv.py [hunk 9]
```diff
@@ -135,24 +168,31 @@ def test_moment_generating_functions():
...
+    assert sellam_mgf.diff(t).subs(
```
Comment (sidhantnagpal) [inline sympy/stats/tests/test_discrete_rv.py]:
```suggestion
    assert skellam_mgf.diff(t).subs(
```

File: sympy/stats/tests/test_discrete_rv.py [hunk 10]
```diff
@@ -172,12 +213,13 @@ def test_conditional():
     assert P(X < 2, Eq(X, 2)) == 0
     assert P(X > 2, Eq(X, 3)) == 1
 
+
 def test_product_spaces():
     X1 = Geometric('X1', S(1)/2)
     X2 = Geometric('X2', S(1)/3)
     assert str(P(X1 + X2 < 3, evaluate=False)) == """Sum(Piecewise((2**(X2 - n - 2)*(2/3)**(X2 - 1)/6, """\
```
Comment (sidhantnagpal) [inline sympy/stats/tests/test_discrete_rv.py]: This could just be:
 ```py
assert P(X1 + X2 < 3, evaluate=False).dummy_eq(Sum(Piecewise(
    (2**(1 - X2)*2**(X2 - n - 2)/4, (-X2 + n + 3 >= 1) & 
    (-X2 + n + 3 < oo)), (0, True)), (X2, 1, oo), (n, -oo, -1)))
```

File: sympy/stats/tests/test_discrete_rv.py [hunk 11]
```diff
@@ -172,12 +213,13 @@ def test_conditional():
...
-    + """(-X2 + n + 3 >= 1) & (-X2 + n + 3 < oo)), (0, True)), (X2, 1, oo), (n, -oo, -1))"""
+        + """(-X2 + n + 3 >= 1) & (-X2 + n + 3 < oo)), (0, True)), (X2, 1, oo), (n, -oo, -1))"""
     assert str(P(X1 + X2 > 3)) == """Sum(Piecewise((2**(X2 - n - 2)*(2/3)**(X2 - 1)/6, """ +\
         """(-X2 + n + 3 >= 1) & (-X2 + n + 3 < oo)), (0, True)), (X2, 1, oo), (n, 1, oo))"""
     assert str(P(Eq(X1 + X2, 3))) == """Sum(Piecewise((2**(X2 - 2)*(2/3)**(X2 - 1)/6, """ +\
-    """X2 <= 2), (0, True)), (X2, 1, oo))"""
+        """X2 <= 2), (0, True)), (X2, 1, oo))"""
```
Comment (sidhantnagpal) [inline sympy/stats/tests/test_discrete_rv.py]: Similar to https://github.com/sympy/sympy/pull/17099#discussion_r298675438.

File: sympy/stats/tests/test_discrete_rv.py [hunk 12]
```diff
@@ -109,7 +140,8 @@ def test_cf(dist, support_lower_limit, support_upper_limit):
 
         # second function is the Fourier transform of the density function
         f = lambdify([x, t], pdf(x)*exp(I*x*t), 'mpmath')
-        cf2 = lambda t: mpmath.nsum(lambda x: f(x, t), [support_lower_limit, support_upper_limit], maxdegree=10)
+        def cf2(t): return mpmath.nsum(lambda x: f(x, t), [
```
Comment (sidhantnagpal) [inline sympy/stats/tests/test_discrete_rv.py]: No need to modify this. It is fine that you split the statement.
```suggestion
        cf2 = lambda t: mpmath.nsum(lambda x: f(x, t), [
```

File: sympy/stats/drv_types.py [hunk 13]
```diff
@@ -3,10 +3,10 @@
 Contains
 ========
 Geometric
-Poisson
```
Comment (Upabjojr) [inline sympy/stats/drv_types.py]: why did you remove this line?

File: sympy/stats/tests/test_discrete_rv.py [hunk 14]
```diff
@@ -1,15 +1,19 @@
-from sympy import S, Sum, I, lambdify, re, im, log, simplify, zeta, pi
-from sympy.abc import x
```
Comment (Upabjojr) [inline sympy/stats/tests/test_discrete_rv.py]: why?

File: sympy/stats/tests/test_discrete_rv.py [hunk 15]
```diff
@@ -18,6 +22,7 @@ def test_PoissonDistribution():
     assert p.expectation(x, x) == l
     assert p.expectation(x**2, x) - p.expectation(x, x)**2 == l
 
+
```
Comment (Upabjojr) [inline sympy/stats/tests/test_discrete_rv.py]: extra empty line

**Comment by Upabjojr** [comment]
There are lots of extra empty lines being added. Can you remove them?

File: sympy/stats/drv_types.py (hunk 13 above)
Comment (sidhantnagpal) [reply sympy/stats/drv_types.py]: It already existed at L9.

File: sympy/stats/tests/test_discrete_rv.py (hunk 14 above)
Comment (ritesh99rakesh) [reply sympy/stats/tests/test_discrete_rv.py]: `from sympy.abc import x` is slower than `x = Symbol('x')`

**Comment by ritesh99rakesh** [comment]
> There are lots of extra empty lines being added. Can you remove them?

@Upabjojr I recently installed `pep8` formatter and I tried to format the file using that. If the changes does not seem right, I will remove them.

**Comment by Upabjojr** [comment]
> 
> 
> > There are lots of extra empty lines being added. Can you remove them?
> 
> @Upabjojr I recently installed `pep8` formatter and I tried to format the file using that. If the changes does not seem right, I will remove them.

PEP8 is not really that uniquely defined. Some decisions are still up to users.

**Comment by Upabjojr** [comment]
> 
> 
> > There are lots of extra empty lines being added. Can you remove them?
> 
> @Upabjojr I recently installed `pep8` formatter and I tried to format the file using that. If the changes does not seem right, I will remove them.

Sorry, I misinterpreted the indentation level.

**Comment by Upabjojr** [comment]
@ritesh99rakesh can you address the issues by @sidhantnagpal ?

File: sympy/stats/tests/test_discrete_rv.py (hunk 10 above)
Comment (ritesh99rakesh) [reply sympy/stats/tests/test_discrete_rv.py]: This is giving `false` as the answer. Could you please guide me on how I can make it work.

Comment (oscarbenjamin) [reply sympy/stats/tests/test_discrete_rv.py]: It's giving False because the expression there is not the same as the one returned (or the one already in the diff here).

The point is not that you should copy the exact line of code here but that you should use dummy_eq rather than string comparison.

Comment (ritesh99rakesh) [reply sympy/stats/tests/test_discrete_rv.py]: @oscarbenjamin I tried multiple ways, but not getting `True` for `dummy_eq`. Is there some way I could get this?

Comment (oscarbenjamin) [reply sympy/stats/tests/test_discrete_rv.py]: I just tried this to see what the expression is and I get
```julia
In [10]: from sympy.stats import *                                                                                                

In [11]:     X1 = Geometric('X1', S(1)/2) 
    ...:     X2 = Geometric('X2', S(1)/3)                                                                                         

In [12]: res = P(X1 + X2 < 3, evaluate=False)                                                                                     

In [13]: res                                                                                                                      
Out[13]: 
  -1     ∞                                                                 
______ ______                                                              
╲      ╲                                                                   
 ╲      ╲                                                                  
  ╲      ╲    ⎧ X₂ - n - 2    X₂ - 1                                       
   ╲      ╲   ⎪2          ⋅2/3                                             
    ╲      ╲  ⎪─────────────────────  for -X₂ + n + 3 ≥ 1 ∧ -X₂ + n + 3 < ∞
    ╱      ╱  ⎨          6                                                 
   ╱      ╱   ⎪                                                            
  ╱      ╱    ⎪          0                          otherwise              
 ╱      ╱     ⎩                                                            
╱      ╱                                                                   
‾‾‾‾‾‾ ‾‾‾‾‾‾                                                              
False  X₂ = 1
```
which looks like gibberish.

Why is the outer sum going from False to -1 and what does that even mean?
```
In [14]: repr(res)                                                                                                                
Out[14]: 'Sum(Piecewise((2**(X2 - n - 2)*(2/3)**(X2 - 1)/6, (-X2 + n + 3 >= 1) & (-X2 + n + 3 < oo)), (0, True)), (X2, 1, oo), (n, -oo, -1))'
```
You'll need to make sure to change 2/3 to S(2)/3 if using the repr above:
```
In [63]: expected = Sum(Piecewise((2**(X2 - n - 2)*(S(2)/3)**(X2 - 1)/6, (-X2 + n + 3 >= 1) & (-X2 + n + 3 < oo)), (0, True)), (X2
    ...: , 1, oo), (n, -oo, -1))
```

The reason dummy_eq doesn't work is maybe because X2 isn't a Dummy symbol. In fact in the two expressions you are comparing one has X2 as a Symbol and the other as a RandomSymbol.
```
In [62]: res.replace(X2.symbol, Dummy(integer=True)).dummy_eq(expected.replace(X2, Dummy(integer=True)))                          
Out[62]: True
```
(integer=True is needed to eliminate the < oo clause.)

I'm not sure that the output you have here is sensible though...

Comment (ritesh99rakesh) [reply sympy/stats/tests/test_discrete_rv.py]: @oscarbenjamin thank you for the insights. Actually this is not my code. I thought of doing some refactoring along with the my code. I will see the result and open another PR for this.

Comment (ritesh99rakesh) [reply sympy/stats/tests/test_discrete_rv.py]: @sidhantnagpal All other reviews have been resolved. Only this one is left. I will open another PR for this. Please consider merging if rest looks good.
